### PR TITLE
Replace entity spawning abstraction

### PR DESCRIPTION
**Pull request by benfrankel**
Fixes https://github.com/TheBevyFlock/bevy_quickstart/issues/223.
Part of https://github.com/TheBevyFlock/bevy_quickstart/issues/203.

Will need to update the design doc as well.

### CODE REVIEW

**Review by MiniaczQ — APPROVED** [review]
pattern looks good, i can look for nits tomorrow if this is still up

**Review by janhohenheim — CHANGES_REQUESTED** [review]
I'm unhappy about the definition site of the construction functions, as I think this kind of function will be very surprising to a newcomer. The solution probably lies in finding a good naming convention, but I can't think of one.

File: src/screens/playing.rs
Comment (janhohenheim) [inline src/screens/playing.rs:36]:
```suggestion
            // Spawn our playing screen UI like HUDs here.
```

`TODO` sounds like something we forgot to add to the template.

File: src/ui/widget.rs
Comment (janhohenheim) [inline src/ui/widget.rs:39]: Maybe rename this to `ButtonWidget` to avoid naming conflicts with `Bevy`.
Also, does this need to be `pub`?

File: src/spawn.rs
Comment (janhohenheim) [inline src/spawn.rs:26]: The distinction between `spawn_with` and `spawn_fn` is currently probably impossible to grasp from the naming alone. I suggest renaming this to `insert_with`.

File: src/spawn.rs
Comment (janhohenheim) [inline src/spawn.rs:159]: Why can't this already have the signature `fn spawn(commands: Commands)`? We don't need to be object safe.

File: src/screens/playing.rs
Comment (janhohenheim) [inline src/screens/playing.rs:33]: It seems very confusing from a first glance when one should use `add_fn` and when `spawn_fn`. If this is confusing to me as a reviewer, I imagine it would be really really confusing to a newcomer.

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: src/ui/widget.rs
Comment (benfrankel) [reply src/ui/widget.rs:39]: The naming conflict shouldn't be an issue since the intended use is `widget::Button`. Unless it's considered that it's an issue regardless. In my own template I have this as `widget::MenuButton` to avoid the conflict.

It needs to be `pub` because you might not be using the `Widgets` extension trait, which only provides spawning functionality, not adding functionality (which you could use to "inherit" from `Button`).

File: src/screens/credits.rs
Comment (janhohenheim) [inline src/screens/credits.rs:15]: Calling `spawn()` here and in similar places *looks* like a bug. I know it isn't, but it very very much looks like we accidentally just called the spawn method.

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: src/spawn.rs
Comment (benfrankel) [reply src/spawn.rs:26]: These both spawn a new entity, though. The difference is whether we're passing in data.

Comment (janhohenheim) [reply src/spawn.rs:26]: Oh wait, you're right. Can a system implement `EntityCommand` so that we can remove this function?

File: src/screens/credits.rs
Comment (benfrankel) [reply src/screens/credits.rs:15]: Well, originally I was going to leave the extension trait that adds this out of the PR, in which case you would have to write a little `fn spawn_credits_screen` system every time. Which isn't too bad. I'm fine with removing `SpawnSystemExt`.

File: src/spawn.rs
Comment (benfrankel) [reply src/spawn.rs:26]: Due to the orphan rule, that would have to be done upstream. But it would be nice if so.

Comment (janhohenheim) [reply src/spawn.rs:26]: Fair enough. We still need a better name though.

File: src/screens/credits.rs
Comment (janhohenheim) [reply src/screens/credits.rs:15]: I think that would help reduce the complexity a bit.

File: src/screens/playing.rs
Comment (benfrankel) [reply src/screens/playing.rs]: Done: https://github.com/TheBevyFlock/bevy_quickstart/pull/231/commits/f891472c0e0eeeeb1ad89d722ca4eb1a7ff76440

File: src/spawn.rs
Comment (janhohenheim) [reply src/spawn.rs:26]: Actually, in this case you *can* bypass the orphan rule by wrapping the system in a struct and then implementing traits on that struct. Not an option for the template because that is severe type magic, but just wanted to let you know.

**Comment by janhohenheim** [comment]
@alice-i-cecile could you also take a look at this?

File: src/screens/credits.rs
Comment (benfrankel) [reply src/screens/credits.rs]: Removed `SpawnSystemExt`.

File: src/screens/playing.rs
Comment (benfrankel) [reply src/screens/playing.rs:33]: An explanation in the design patterns doc should mitigate that a bit, but I'm open to bikeshedding.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I really don't like this, especially in learning material. There's a *ton* of complexity and type magic here, and it makes simple tasks appear esoteric and difficult. Boilerplate is better than complexity for this.

I understand the desire to improve things here, but I would be strongly opposed to endorsing code with these patterns as official learning materials.

**Comment by janhohenheim** [comment]
@alice-i-cecile I'm tending the same way. I feel a bit bad since we both gave a thumbs-up for it in https://github.com/TheBevyFlock/bevy_quickstart/issues/223, but seeing it in practice is just too "scary". I've been thinking about ways to improve the design and make it more approachable, but I'm really struggling with that.

It would be easier if

- this was an upstreamed API, and so was able to be simplified due to no orphan rule, or
- this was a library, and so a user is opting into it and ready to deal with the new concepts.

For the moment, I'd instead go for the simpler route of replacing observers with custom commands.

**Comment by benfrankel** [comment]
Custom commands means having to extract system parameters from `World`, which is even worse in that regard, no?

**Comment by janhohenheim** [comment]
I don't like extracting stuff manually from `World`, but at least that is "standard Bevy", for better or worse.
Using an `Entity` as the first param is a good analogy to observers, agreed. I personally don't mind it, but I think the Bevy maintainers will disagree with introducing this as an abstraction.

My main issues are more:
- Using the provided abstraction requires active learning, which I don't think is as bad for the other abstractions we use. You can take a single look at `fn plugin` or at `commands.button` and you know how to use it. The same cannot be said for things like the interplay / differences between `spawn_with`, `spawn_fn`, `add_fn`.
- `spawn.rs` feels scary enough that for newcomers it might as well be a black box imo.

As for `Commands` vs `Observer`s: I only recommend using commands because that was @alice-i-cecile's feedback. I personally would prefer an observer because of the ergonomics, even when that means a layer of indirection. My opinion on this is not strong however and I will accept either implementation. I can see both points.

**Comment by benfrankel** [comment]
For the record, custom commands still requires `spawn_with` and extension traits. It would only mean removing `spawn_fn` and `add_fn` and extracting everything from `World` instead of using system parameters.

**Comment by janhohenheim** [comment]
@benfrankel I think that could work, in terms of not introducing too many new abstractions / concepts.

Worst case we can still use the crappy variant that does not return an `Entity` to the caller. It's suboptimal, but it's not our fault that this is fundamentally missing in Bevy and it _might_ be fixed by BSN.

**Comment by benfrankel** [comment]
Done. Usability is much worse without system param support IMO but it's w/e.

File: src/spawn.rs
Comment (janhohenheim) [inline src/spawn.rs:64]:
```suggestion
    // Workaround for <https://github.com/bevyengine/bevy/issues/14278>.
```
Link is currently broken because of the period.

File: src/spawn.rs
Comment (janhohenheim) [inline src/spawn.rs:41]:
```suggestion
    // Workaround for <https://github.com/bevyengine/bevy/issues/14231#issuecomment-2216321086>.
```
Link does not point to the comment due to the period

File: src/spawn.rs
Comment (janhohenheim) [inline src/spawn.rs:18]:
```suggestion
    // Workaround for <https://github.com/bevyengine/bevy/issues/14231#issuecomment-2216321086>.
```
Link does not point to the comment due to the period

File: src/demo/player.rs
Comment (janhohenheim) [inline src/demo/player.rs:27]: Could we rename these to `construct_player` and `construct_level` now? Seeing as the underlying pattern has shifted.
`children.spawn_with(construct_player);` sounds completely reasonable to be.

File: src/demo/player.rs
Comment (janhohenheim) [inline src/demo/player.rs:26]:
```suggestion
/// Construct a player entity by inserting the relevant components.
```

**Review by janhohenheim — CHANGES_REQUESTED** [review]
I think this version is *way* clearer for an outsider. Other than some minor issues I noted, I'd merge this. 
But I also want to hear @alice-i-cecile's opinion again, as this is controversial enough imo. 
I want to stress that the underlying spawning API is missing from Bevy, so we *will* either have a somewhat complex, but decently usable pattern, or a very simple pattern that will break down when the user experiments with it.
I think this PR strikes a good balance. I'd prefer to ship *some* way of spawning stuff while returning `Entity` or `EntityCommands` or similar as, in contrast to many other tradeoffs of the template, this is *not* something where a user can just switch to a popular community plugin when they are running into trouble.

File: src/demo/level.rs
Comment (janhohenheim) [inline src/demo/level.rs:16]: Same as for `fn player`

File: src/ui/widget.rs
Comment (janhohenheim) [inline src/ui/widget.rs:211]: same as `fn player`

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: src/spawn.rs
Comment (benfrankel) [reply src/spawn.rs:64]: Where does the link not work? It works for me in VS Code like this.

File: src/demo/player.rs
Comment (benfrankel) [reply src/demo/player.rs:27]: I still think that's more confusing (verbosity aside).

File: src/demo/player.rs
Comment (benfrankel) [reply src/demo/player.rs:26]: This could add children / do other things as well.

Comment (janhohenheim) [reply src/demo/player.rs:26]: I would like to guide users more on what "construct" means in this context and why there is a `Entity` as a parameter.

File: src/spawn.rs
Comment (janhohenheim) [reply src/spawn.rs:64]: Clicking it on my Firefox redirects me to this page: https://github.com/bevyengine/bevy/issues/14278.?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen
I think your VS Code is just smart enough to figure out that the dot is not actually part of the URL, while my browser / GitHub assumes it is.

File: src/demo/player.rs
Comment (janhohenheim) [reply src/demo/player.rs:27]: I think we are at an impasse with the naming here. Requesting a third opinion from anyone, I'll gladly accept the majority vote.

File: src/spawn.rs
Comment (alice-i-cecile) [inline src/spawn.rs:18]: IMO this should just be fixed upstream.

**Comment by alice-i-cecile** [comment]
I think this is an improvement, but I'm still not happy with the level of abstraction / ad-hoc tools here. I think this should be merged, but the linked bug / limitation should be fixed upstream (ping me for reviews!), and I'm still reluctant to officially recommend this template until this is substantially simpler.

If you feel that the limitations of just doing things the dumb way are unacceptable, I'm fine to hold off on endorsing this until the BSN work is shipped.

**Comment by janhohenheim** [comment]
Closing this for now. Sorry again for the whole work that is not merged. Let's revisit this after some improvements have been implemented upstream.

**Comment by benfrankel** [comment]
All good, it was mostly copy/paste and some mechanical refactoring.

**Pull request closed without merging**
